tutorials/en/vectara-advanced-app-tutorial-showcase-the-creation-of-vectara-app-in-legal-or-customer-support-use-case.mdx

```diff
@@ -0,0 +1,671 @@
+
+```python
+    def upload_file(self, customer_id: int, corpus_id: int, idx_address: str, uploaded_file, file_title: str) -> Tuple[requests.Response, bool]:
+            """Uploads a file to the corpus."""
+            # Determine the MIME type based on the file extension
+            extension_to_mime_type = {
+                '.txt': 'text/plain',
+                '.pdf': 'application/pdf',
+                '.doc': 'application/msword',
+                '.docx': 'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
+                # ... add more mappings as needed
+            }
+            file_extension = os.path.splitext(uploaded_file.name)[-1]
+            mime_type = extension_to_mime_type.get(file_extension, 'application/octet-stream')
+
+            post_headers = {
+                "Authorization": f"Bearer {self.jwt_token}"
+            }
+            
+            try:
+                file = uploaded_file.read()  
+                files = {"file": (file_title, file, mime_type)}
+                response = requests.post(
+                    f"https://{idx_address}/v1/upload?c={customer_id}&o={corpus_id}",
+                    files=files,
+                    headers=post_headers
+                )
+            
+                if response.status_code != 200:
+                    logging.error("REST upload failed with code %d, reason %s, text %s",
+                                response.status_code,
+                                response.reason,
+                                response.text)
+                    return response, False
+                return response, True
+            except Exception as e:
+                logging.error("An error occurred while uploading the file: %s", str(e))
+                return None, False
+ 
+        ...
+```
+
+The `upload_file` method is our gateway to send documents to Vectara's corpus. It's designed to handle the file upload, ensuring the document finds its place in the corpus for later retrieval.
+
+##### **Automated MIME Type Detection** 🔍
+
+```python
+        # Determine the MIME type based on the file extension
+        extension_to_mime_type = {
+            '.txt': 'text/plain',
+            '.pdf': 'application/pdf',
+            '.doc': 'application/msword',
+            '.docx': 'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
+            # ... add more mappings as needed
+        }
+        file_extension = os.path.splitext(uploaded_file.name)[-1]
+        mime_type = extension_to_mime_type.get(file_extension, 'application/octet-stream')
+```
+
+In the snippet above, we manually map file extensions to their respective MIME types. This is crucial for the Vectara platform to understand the nature of the file and process it accordingly. However, it's worth noting that the `mimetypes` library could also be used to dynamically determine the MIME type of the file being uploaded, as shown below:
```

Comment:
Same comment as above - why do we need this? Vectara identifies if it's PDF or DOC or PPT once the file is uploaded.